session.go

```diff
@@ -0,0 +1,202 @@
+	// SDK calls needed for sending relays.
+	res, err := s.PoktNodeSessionFetcher.GetSession(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+
+	return res.Session, nil
+}
+
+// NewPoktNodeSessionFetcher returns the default implementation of the PoktNodeSessionFetcher interface.
+// It connects to a POKT full node through the session module's query client to get session data.
+func NewPoktNodeSessionFetcher(grpcConn grpc.ClientConn) PoktNodeSessionFetcher {
+	return sessiontypes.NewQueryClient(grpcConn)
+}
+
+// PoktNodeSessionFetcher is used by the SessionClient to fetch sessions using poktroll request/response types.
+//
+// Most users can rely on the default implementation provided by NewPoktNodeSessionFetcher function.
+// A custom implementation of this interface can be used to gain more granular control over the interactions
+// of the SessionClient with the POKT full node.
+type PoktNodeSessionFetcher interface {
+	GetSession(
+		context.Context,
+		*sessiontypes.QueryGetSessionRequest,
+		...grpcoptions.CallOption,
+	) (*sessiontypes.QueryGetSessionResponse, error)
+}
+
+// TODO_IMPROVE: add a `FilterSupplier` method to drop a supplier from consideration when sending a relay.
+// TODO_IMPROVE: add a `FilterEndpoint` method to drop an endpoint from consideration when sending a relay.
+// TODO_IMPROVE: add an OrderEndpoints to allow ordering the list of available, i.e. not filtered out, endpoints.
+// The `SelectedEndpoint` method (and potentially a `SelectedEndpoints` method) should then return the ordered list of endpoints for the target service.
+//
+// FilteredSession wraps a Session, allowing node selection, currently through directly adding endpoints, and potentially through filtering/ordering in future versions.
+// This is needed so functions that enable sending relays can be provided with a struct that contains both session data and the endpoint(s) selected for receiving relays.
+type FilteredSession struct {
+	*sessiontypes.Session
+
+	// selectedEndpoints is the set of a specific service's endpoints, from the underlying session, that are selected by the user of FilteredNodes.
+	// This selection can be for any purpose, e.g. sending relays.
+	// The set of selected endpoints is stored as a map of supplier addresses to a list of endpoints.
+	selectedEndpoints map[string][]*sharedtypes.SupplierEndpoint
+
+	// selectedServiceId stores the ID of the service selected by the user.
+	selectedServiceId string
+}
+
+// TODO_DISCUSS: Consider using a custom type, defined as a string, as supplier address.
+// This can help enforce type safety by requiring explict type casting of a string before it can be used as a SupplierAddress.
+//
+// ServiceEndpoints returns the supplier endpoints assigned to a session for the given service id.
+//
+// The returned value is a map of supplierId to the list of endpoints.
+func (f *FilteredSession) ServiceEndpoints(
+	serviceId string,
```

Comment:
Thank you for the review: based on my reading of the `Session` struct, it can contain multiple suppliers, and each supplier could contain multiple services, hence the need for specifying a `serviceId` (which I agree looks redundant, but may actually be needed due to how the Supplier struct works). Is this accurate?

Reply:
We are sure that the suppliers contained in the `Session` provide the service that the `Session` is about. It's just that the on-chain part is not removing the unrelated services from those Suppliers.

Which means that all `Supplier`s in the `Session` have at least 1 endpoint related to the `serviceId`.

I think this method is a good place to remove the unrelated endpoints and return only the relevant ones.